libraries/aws/shadow/src/shadow.c

```diff
@@ -0,0 +1,550 @@
+/* coverity[misra_c_2012_rule_5_4_violation] */
+#define SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DELTA                  ( SHADOW_TOPIC_OPERATION_LENGTH_UPDATE + SHADOW_TOPIC_SUFFIX_LENGTH_DELTA )
+
+/**
+ * @brief The length of "/shadow/get/accepted".
+ */
+
+/* MISRA Rule 5.4 flags the following macro's name as ambiguous from the
+ * one without _ACCEPTED suffix. This rule is suppressed for naming consistency with
+ * other Shadow header field and value string and length macros in this file.*/
+/* coverity[misra_c_2012_rule_5_4_violation] */
+#define SHADOW_TOPIC_OPERATION_LENGTH_GET_ACCEPTED                  ( SHADOW_TOPIC_OPERATION_LENGTH_GET + SHADOW_TOPIC_SUFFIX_LENGTH_ACCEPTED )
+
+/**
+ * @brief The length of "/shadow/get/rejected".
+ */
+
+/* MISRA Rule 5.4 flags the following macro's name as ambiguous from the
+ * one without _REJECTED suffix. This rule is suppressed for naming consistency with
+ * other Shadow header field and value string and length macros in this file.*/
+/* coverity[misra_c_2012_rule_5_4_violation] */
+#define SHADOW_TOPIC_OPERATION_LENGTH_GET_REJECTED                  ( SHADOW_TOPIC_OPERATION_LENGTH_GET + SHADOW_TOPIC_SUFFIX_LENGTH_REJECTED )
+
+/**
+ * @brief The length of "/shadow/get/accepted".
+ */
+
+/* MISRA Rule 5.4 flags the following macro's name as ambiguous from the
+ * one without _ACCEPTED suffix. This rule is suppressed for naming consistency with
+ * other Shadow header field and value string and length macros in this file.*/
+/* coverity[misra_c_2012_rule_5_4_violation] */
+#define SHADOW_TOPIC_OPERATION_LENGTH_DELETE_ACCEPTED               ( SHADOW_TOPIC_OPERATION_LENGTH_DELETE + SHADOW_TOPIC_SUFFIX_LENGTH_ACCEPTED )
+
+/**
+ * @brief The length of "/shadow/delete/rejected".
+ */
+
+/* MISRA Rule 5.4 flags the following macro's name as ambiguous from the
+ * one without _REJECTED suffix. This rule is suppressed for naming consistency with
+ * other Shadow header field and value string and length macros in this file.*/
+/* coverity[misra_c_2012_rule_5_4_violation] */
+#define SHADOW_TOPIC_OPERATION_LENGTH_DELETE_REJECTED               ( SHADOW_TOPIC_OPERATION_LENGTH_DELETE + SHADOW_TOPIC_SUFFIX_LENGTH_REJECTED )
+
+/**
+ * @brief Determine if the string contains the substring.
+ *
+ * @param[in] pString Pointer to the string.
+ * @param[in] stringLength Length of pString.
+ * @param[in] pSubString Pointer to the substring.
+ * @param[in] subStringLength Length of pSubString.
+ *
+ * @return Return SHADOW_STATUS_SUCCESS if it contains;
+ *         return SHADOW_STATUS_FAIL if not.
+ */
+static ShadowStatus_t containsSubString( const char * pString,
+                                         uint16_t stringLength,
+                                         const char * pSubString,
+                                         uint16_t subStringLength );
+/**
+ * @brief Extract the Thing Name from a string.
+ *
+ * @param[in] pString Pointer to the string.
```

Comment:
We should specify that this parameter must already be pointing to the first character of the thing name

Reply:
pString is the string to be extract, pThingName is the output pointer that points to the thing name in the topic.

If we remove the pThingName parameter, it still requires the caller which called extractThingName to assign  pThingName to pString when get success results. Which is the same thing we are doing in extractThingName()
( This is because Shadow_MatchTopic outputs pThingName)

Please let me know if you still suggest to make this change

Reply:
Please update either the comment or the function name, since it isn't immediately clear from the name of the function that `pString` should already be pointing to the first character of the string name. Just going by the name of the function, I'd expect it to search the string for the start of the thing name, instead of just determining the length of the thing name.

Also, regarding `pThingName`, including it as a parameter requires an extra line by the caller in the `extractThingName ` function call, while removing it would require an extra line for the caller to set the thing name, so the net result is the same. I suggested removing it since it looks like this function mainly just determines the length of the thing name, and assigning an input to an output unchanged seems unnecessary for this function to do

Reply:
This assignment can be done in `Shadow_MatchTopic`:


```c
    if( shadowStatus == SHADOW_SUCCESS )
    {
        /* Extract thing name. */
        shadowStatus = extractThingName( & ( pTopic[ consumedTopicLength ] ),
                                         topicLength - consumedTopicLength,
                                         pThingNameLength );
        if( shadowStatus == SHADOW_SUCCESS )
        {
            /* Update the out parameter if we successfully extracted the thing name. */
            pThingName = & ( pTopic[ consumedTopicLength ] );

            consumedTopicLength += * pThingNameLength;

            /* If no more topic string is left to parse, fail. */
            if( consumedTopicLength >= topicLength )
            {
                shadowStatus = SHADOW_SHADOW_MESSAGE_TYPE_PARSE_FAILED;
                LogDebug( ( "Not related to Shadow, shadow message type is not in pTopic %s, failed to parse shadow message type", pTopic ) );
            }
        }
        else
        {
            LogDebug( ( "Not related to Shadow, failed to parse thing name in pTopic %s", pTopic ) );
        }
    }
```

Reply:
fixed it in commit [8da68c5](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1081/commits/8da68c5f3ccc011a9ab8f58349cf6273c42ec19c)

by removing pThingName for extractThingName, and assign it in Shadow_MatchTopic

resolving this comment

feel free to let me know if it needs further changes, thank you